### ISSUE TITLE
Some expressions take an enormous amount of time to parse

### ISSUE THREAD

**Issue by RoystonS** [b3b00/csly#254]
Hi there,

I've just upgraded from sly 2.7.0.3 to 2.7.0.5 and switched from using laborious manually-driven precedence to the lovely new built-in precedence support. However, a couple of my expressions work but are _incredibly_ slow.

Here's an example piece of code for a project I maintain.  This commit moves me from 2.7.0.3 to 2.7.0.5, introduces bitwise '&' and '|' operators and moves to using `[Operand]`, `[Infix]` and `[Prefix]` attributes: https://github.com/RoystonS/BetterBravoLights/commit/6d066a9f986da911992bc5978946550d7546a187

All of the tests pass, but two in particular are now _incredibly_ slow.  I'm not sure what's going on - I'd expect them either to fail with some sort of stack overflow exception due to excessive recursion, or pass very quickly.

![image](https://user-images.githubusercontent.com/19773/140650323-a41e8bb8-6ba4-4f3c-b95c-d35974b5a3d5.png)

To reproduce, grab the above project at that commit, built it in VS2019 and run the unit tests.

The particular case that's problematic is one that mixes unary minus with parenthetic grouping, i.e. `-(1+2)`.

**Comment by b3b00** [comment]
hello @RoystonS , I will look at it real quick. Were test cases already slow prior to 2.7.0.5 ?

**Comment by b3b00** [comment]
hello, I've cloned your repository (using branch use-new-sly-expressions) I have not FlightSim so I can not build it 🤕 
So I've extracted your parser and lexer and created a simple console application to test it (see branch https://github.com/b3b00/csly/tree/bugfix/issue254 )
I do not see any performance issue. Are you sure that it does not come from your code when avaluating the produced AST ? this part I can not test due to flightsim lack.
nevertheless using these test cases form ParserTests.LiteralNumericExpressionsEvaluateCorrectly all parses fail with an unexpected EOS error. I've got to investigate this.

**Comment by b3b00** [comment]
@RoystonS sorry all is fine with your parser : all test case ar ok. I 've just messed with branches :) 
Indeed I observe 2 test cases way slower thant other (see below) But not in the proportion that you describe.

some time measurement :

```
test 1 + 2 * 3 : True : 53 ms
test 2 - 3 / 4 : True : 7 ms
test -3 - -4 : True : 3 ms
test -3--4 : True : 4 ms
test -3+-4 : True : 3 ms
test -(1+2) : True : 387 ms
test -(1+2 * 3) : True : 382 ms
test 3 * -2 : True : 4 ms
test 9 & 8 : True : 2 ms
test 7 & 8 : True : 4 ms
test 8 & 8 : True : 1 ms
``` 

I will follow with some profiling on -(1+2)  and -(1+2 * 3) expressions

**Comment by RoystonS** [comment]
Hi @b3b00. Sorry, completely forgot you wouldn't be able to build it without the flight sim part!
Sounds like you've managed to pick the relevant parts apart though - nicely done!

The evaluation of my AST hasn't changed between the sly 2.7.0.3 and 2.7.0.5 branches so I don't think the evaluation time is a hugely significant factor.

Does look like you've managed to reproduce the problem: you've got 2 cases that take ~100 times longer than the others. That roughly matches mine, where most of my tests are 300ms but those 2 cases take 40-50 seconds.  Not sure why my tests are all so slow. I'll look at the branch you've produced.

**Comment by b3b00** [comment]
Hi @RoystonS ,

when using CSLY Operations , CSLY generates all the needed grammar rules to manage precedence. In your case you have many different precedneces and so the full grammar is quite big (see below). this causes a lot of bactracking slowing the parse.  

```
numeric_literal : LVAR 
numeric_literal : SIMVAR 

group : LPAREN MSFSExpressionParser_expressions RPAREN 

logical_literal : OFF 
logical_literal : ON 

primary : HEX_NUMBER 
primary : DECIMAL_NUMBER 

MSFSExpressionParser_operand : numeric_literal 
MSFSExpressionParser_operand : numeric_literal 
MSFSExpressionParser_operand : group 
MSFSExpressionParser_operand : logical_literal 
MSFSExpressionParser_operand : primary 
MSFSExpressionParser_operand : primary 

expr_6_LOGICAL_OR : expr_7_LOGICAL_AND LOGICAL_OR expr_6_LOGICAL_OR 
expr_6_LOGICAL_OR : expr_7_LOGICAL_AND 

expr_7_LOGICAL_AND : expr_8_BITWISE_OR LOGICAL_AND expr_7_LOGICAL_AND 
expr_7_LOGICAL_AND : expr_8_BITWISE_OR 

expr_8_BITWISE_OR : expr_10_BITWISE_AND BITWISE_OR expr_8_BITWISE_OR 
expr_8_BITWISE_OR : expr_10_BITWISE_AND 

expr_10_BITWISE_AND : expr_11_NOT BITWISE_AND expr_10_BITWISE_AND 
expr_10_BITWISE_AND : expr_11_NOT 

expr_11_NOT : NOT expr_12_COMPARISON 
expr_11_NOT : expr_12_COMPARISON 

expr_12_COMPARISON : expr_14_PLUS_MINUS COMPARISON expr_12_COMPARISON 
expr_12_COMPARISON : expr_14_PLUS_MINUS 

expr_14_PLUS_MINUS : expr_15_TIMES_DIVIDE [ PLUS | MINUS ] expr_14_PLUS_MINUS 
expr_14_PLUS_MINUS : expr_15_TIMES_DIVIDE 

expr_15_TIMES_DIVIDE : expr_17_MINUS [ TIMES | DIVIDE ] expr_15_TIMES_DIVIDE 
expr_15_TIMES_DIVIDE : expr_17_MINUS 

expr_17_MINUS : MINUS MSFSExpressionParser_operand 
expr_17_MINUS : MSFSExpressionParser_operand 

MSFSExpressionParser_expressions : expr_6_LOGICAL_OR 

```

**Comment by b3b00** [comment]
@RoystonS I 've compared your parser on main branch with the one using CSLY Operations. It appears that you 've added bitwise operations and removing them from the parser on branch use-new-sly-expressions make the parser perform way better on expressions like -(1+2) or -(1+2 * 3). Of course expressions using bitwise operators will failed 

I have to find a way to make CSLY better manage such grammars with a lot of operator and different precedences.

**Comment by RoystonS** [comment]
Yes, I added bitwise operations on the new branch. That's what led me to try the new operations code because manually managing all the precedence rules was becoming quite onerous.  (I was also seeing some nasty issues with lexing getting confused between `&` and `&&` which coincided with what I thought was unit tests that were hanging, but was probably just the unit tests being very slow.  That's why I thought I'd try the new Operations.)

And yes, the performance of those same expressions in the pre-bitwise grammar was much better.  It felt like the addition of just that one extra rule tipped the performance right over the edge?

I'm going to take a look as to why those tests are taking _such_ a long time for me compared to your runs.

The other thing I noticed whilst switching to using CSLY Operations is that my previous manual-precedence grammar looks to be have been (incorrectly) largely right-to-left associative. Would that also be linked to the need for the extra backtracking?

Interesting that you're classing this as a 'big' grammar. The precedence values I chose to use are mostly from the JavaScript operators list, which is much much bigger: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence#table

The good news is that in my real app the actual _parsing_ doesn't run all that often, whereas the evaluation of the ASTs can run many times per second.

**Comment by RoystonS** [comment]
Ah, I just realised the other thing that changed when I moved from the main branch to the Operations branch. All operands are now treated the same. Previously I had numeric operands and logical operands. Would that have made the parsing more efficient too?

That is, the leaf level of this grammar (at least previously) pretty much always consisted of numeric expressions and the root level is a boolean expression.  So, you can do basic arithmetic on numbers to get numbers, perform comparisons to get booleans out of numbers, and logical operations to get booleans out of booleans. But you can never get numbers from booleans.  I think the new Operations grammar - because it lumps all the operands together - is a slightly more flexible (read: inefficient?) grammar than I had before?

**Comment by b3b00** [comment]
associativity is not the cause of backtracking. it is managed once the syntax tree is built.
CSLY is not designed to parse full featured languages like javascript but rather small DSL like yours, and grammars for such languages are usually quite small.

**Comment by RoystonS** [comment]
I've just tried applying your generated parse productions manually to see if I could use that as a start point for tweaking.
(Modified parser is here: https://github.com/RoystonS/BetterBravoLights/tree/generated-sly-expressions)

Curiously, 4 of my 52 unit tests fail, which I wasn't expecting. It looks like the associativity is different?  Have I screwed something up in the translation? I tried to follow the grammar in your previous generated listing: https://github.com/b3b00/csly/issues/254#issuecomment-963171550

```
ParserSupportsAllArithmeticOperators:
Expected: ((((((1 < 2) AND (1 <= 2)) AND (1 == 2)) AN···
Actual:   ((1 < 2) AND ((1 <= 2) AND ((1 == 2) AND ((···
            ↑ (pos 2)
```

I'd been hoping I could reproduce the generated grammar and then simplify it and improve its performance by keeping the boolean portions of the grammar away from the logical portions.

**Comment by b3b00** [comment]
@RoystonS , CSLY generates a right recursive grammar and so right associativity. Associativity is managed after syntax parsing doing some syntax tree manipulation. So using the generated grammar will produce a right associative  syntax tree and your unit tests will fail.

**Comment by b3b00** [comment]
I've just noticed sommething strange with the generated grammar. You define : 

```c#
[Operand]
[Production("primary: HEX_NUMBER")]

// and 

 [Operand]
 [Production("primary: DECIMAL_NUMBER")]

``` 

this produces 2 grammar rules : 
```
MSFSExpressionParser_operand : primary 
MSFSExpressionParser_operand : primary 
```

This may cause useless backtracking. 
And the same goes for ```MSFSExpressionParser_operand : numeric_literal```

I will look if I can deduplicate this rules. not sure if it will show better perf but worth a try.

**Comment by b3b00** [comment]
Indeed there is a benefice. I 've pushed a fix on branch [bugfix/issue254]( https://github.com/b3b00/csly/tree/bugfix/issue254).
@RoystonS, Could you try to clone the CSLY repo and link your project to it to validate it ?

**Comment by RoystonS** [comment]
Hi @b3b00. Sorry, I've had a huge amount of work commitments over the last few days so haven't been able to look at this.

I've just picked up your [bugfix/issue254](https://github.com/b3b00/csly/tree/bugfix/issue254) branch, run a Release build and ran my tests against it but it doesn't _really_ improve the performance in any noticeable way:
![image](https://user-images.githubusercontent.com/19773/141599506-aa85ba3e-bb40-49db-a256-716a1803e254.png)

Any suggestions - other than to reduce the number of precedences - to work around the performance issue?

**Comment by b3b00** [comment]
may I ask what is your configuration ? cpu memory OS version and framework version (dotnet --info) ?

**Comment by RoystonS** [comment]
Hi. Sure.

I've an AMD Ryzen 5 3600 with 32GB (4x8GB) of 3600MHz RAM. It's fairly decent. It's running a Windows 11 Insiders build (10.0.22000.346) and is fully patched with all the AMD cache latency hot fixes.  I'm not sure how my timings are _so_ much slower than the ones you've posted: the single-thread performance of a 3600 certainly doesn't match up to that of a 5800X, 12th gen Intel or an M1 but it's not terrible.

`dotnet --info` output:
```
.NET SDK (reflecting any global.json):
 Version:   5.0.403
 Commit:    bf9d168ac2

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22000
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.403\

Host (useful for support):
  Version: 5.0.12
  Commit:  7211aa01b3

.NET SDKs installed:
  3.1.415 [C:\Program Files\dotnet\sdk]
  5.0.400 [C:\Program Files\dotnet\sdk]
  5.0.402 [C:\Program Files\dotnet\sdk]
  5.0.403 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.8 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
```

I'm using Visual Studio 2019 Community Edition.

As an experiment I've been trying out some alternative parsers, to see whether it's an issue with my grammar and/or my machine or the way I'm building my library or tests. I don't particularly like the Superpower library's API (it's quite confusing compared to csly's) but it does seem to be running all my unit tests successfully in just over 2 seconds total.  (In particular the two most problematic test cases both run in 8ms with Superpower compared with 60 & 46 seconds with csly.)

<img width="452" alt="image" src="https://user-images.githubusercontent.com/19773/142401975-c3349cd5-ec60-42c7-9c46-c321e9e13d64.png">

Historically, I'm most familiar with Yacc/Bison syntax, which is why I like csly's 'production string' syntax.  I've resisted firing up ANTLR because I don't really want a separate build process for the parser.  But if the grammar I have isn't a good fit for csly, that's absolutely fine: Superpower would give me a way forward.

**Comment by b3b00** [comment]
thanks for your feedback. I have now to think about this perf issue and your gramar is a great basis to look at it.

**Comment by b3b00** [comment]
@RoystonS, using superpower how do you manage left associativity ?
do you have a branch with the superpower grammar ? i wouls like to take it at a basis for performance improvment.

**Comment by b3b00** [comment]
@RoystonS I 've tested some optimization for expression parsin. You can find it on branch https://github.com/b3b00/csly/tree/experiment/optimize-expression-rules
It's still a work in progress. My local tests show a great improvment on your problematic expressions.
But error reporting got some regression in the process.

Could you try it , if you have enough time ?

**Comment by RoystonS** [comment]
Heya. Thanks for doing that. Yep, I'll get that built up and tested, hopefully later today. I'll also get you the other superpower branch to take a look at for comparison.

**Comment by RoystonS** [comment]
When you say "Great improvement", that turns out to be a bit of an understatement!

Before (using your issue254 branch):
<img width="304" alt="image" src="https://user-images.githubusercontent.com/19773/143135326-7cbf392b-a461-43a7-a1a2-7ef6b8597aa8.png">

After:
<img width="299" alt="image" src="https://user-images.githubusercontent.com/19773/143135637-4325ea24-8c5d-485c-85aa-ea855999e2ec.png">

That's a Release build of csly and a Debug build of my project, both using VS2022.

Here's the superpower branch out of interest:
https://github.com/RoystonS/BetterBravoLights/tree/use-superpower-parser
And a test run against that (a slightly different set of tests but most are the same):
<img width="459" alt="image" src="https://user-images.githubusercontent.com/19773/143136152-666abf0a-75fb-491c-a64c-f7efd4cdb13f.png">

So your updated code is now significantly faster even than Superpower. That's an impressive piece of optimisation, sir!

**Comment by b3b00** [comment]
Helo, thanks for your testing. I 've just publish an alpha version : https://www.nuget.org/packages/sly/2.8.0-alpha
I still need some debug cause error reporting is not as accurate as before, I have 1 failing unit test. 

I hope final release wil come real soon.

Thanks for your help improving CSLy performance

**Comment by b3b00** [comment]
I've just published v 2.8.0. could you test it and tell me if all is ok ?

**Comment by RoystonS** [comment]
Yep, 2.8.0 looks absolutely fantastic. Outstanding job, sir - **very** much appreciated!
